### ISSUE TITLE
Rethink data-abstraction

### ISSUE THREAD

**Issue by LeanderK** [pesser/edflow#57]
I don't think the data-abstraction is really useful right now:
- no ability to define train/test/validation splits
- no ability for more "flexible" datasets (for example, I need to change the dataset halfway to a time-series dataset in next project)
- no ability to define more than one datasets for train/test/validation (multiple datasets for training are common in multi-task projects, I need multiple validation/test datasets to evaluate different metrics)

I think it needs to be reworked to be really useful outside the current scope of the project. This introduces more complexity and may not be desired by the authors, but nevertheless should be discussed. 

I think the interpretation of the data-part should be more holistic and therefore not have a single data-set as the fundamental abstraction, but instead a collection.

**Comment by theRealSuperMario** [comment]
I totally agree that the data management situation is not ideal.

1. In theory, splits can be handeled with -e eval1.yaml, -e eval2.yaml. Then each evaluation would be a validation set.
2. I agree that we do not have a nice and clean hook that fits in here.

This is definately something we are working.

**Comment by LeanderK** [comment]
I would be happy to implement something like this. This ticket was meant to start a discussion and plan the feature so that it can be implemented afterward.

> In theory, splits can be handeled with -e eval1.yaml, -e eval2.yaml. Then each evaluation would be a validation set.

This assumes the evaluation is independent, but in my case, the training dataset defines the evaluation datasets (they are different or related versions of the training datasets).

**Comment by LeanderK** [comment]
To be clear, this is a roadblock for me and I can't migrate to edflow until this is not solved. Right now, I have some time until ICLR to invest time into tooling, but afterwards, I am probably busy experimenting @theRealSuperMario .

**Comment by pesser** [comment]
Hi LeanderK,

> I don't think the data-abstraction is really useful right now:
> 
>     * no ability to define train/test/validation splits

This is really left to the dataset implementation. It always has access to the configuration, so you can just specify a split argument there and filter your data based on it. [SubDataset](https://github.com/pesser/edflow/blob/1a2cf357e70be0214c49f6ebf327a363a74a12d7/edflow/data/dataset.py#L522) is handy for that task. Maybe a quick word on the undocumented world our datasets: The dataset specified in the config does not have to be a class, it can simply be a function which gets called with the config. It can then construct a dataset inheriting from `DatasetMixin`.

>     * no ability for more "flexible" datasets (for example, I need to change the dataset halfway to a time-series dataset in next project)

There is currently one main limitation that I can think of: We assume a fixed batch size and compatible shapes within each batch, i.e. each example in a batch must have the same shape. We also have some [helpers](https://github.com/pesser/edflow/blob/1a2cf357e70be0214c49f6ebf327a363a74a12d7/edflow/data/dataset.py#L755) which we use for video datasets. The general idea is then just that `get_example` returns a sequence or pairs or whatever one wants. You can still reshape the sequence to the batch axis later if desired. There are also other operations to concatenate, join or process datasets arbitrarily.

As for switching halfway to a different kind of dataset, I would definitely suggest to split the training into two stages and just resume training on the new dataset in the second stage. Is that feasible or do you think there is a really good reason to change the dataset during training?

>     * no ability to define more than one datasets for train/test/validation (multiple datasets for training are common in multi-task projects, I need multiple validation/test datasets to evaluate different metrics)

This is basically exactly what we had in mind, too: Each evaluation might require a different dataset, and each evaluation has its own config which includes the dataset.

> I think it needs to be reworked to be really useful outside the current scope of the project. This introduces more complexity and may not be desired by the authors, but nevertheless should be discussed.

It needs a lot of work and additional complexity is fine if required.

> I think the interpretation of the data-part should be more holistic and therefore not have a single data-set as the fundamental abstraction, but instead a collection.

Currently, the idea is that each training and each evaluation is as independent as possible from the others. Evaluations run in separate processes, which is on the one hand nice, because (i) they do not interfere with training, (ii) they can be started anytime. The big downside to this is that evaluations only run whenever a checkpoint is written. In theory, one can write checkpoints often and keep only the best according to the evaluation, but for quick experimentation it can be tedious to do so.

I see a couple of directions to improve this situation:
1. Collections of datasets as you mention: This would enable different datasets in a single process, but the only use case I can see is that of evaluating the model. If we go that route, then we basically abandon the idea of separation between training and evaluation and we always have to handle logic for training and evaluation during development. I actually prefer the separation because it allows quicker iteration: I can first write whatever code is necessary to start my training, and then I can still think about evaluation code while it trains. If we mix training and evaluation and I change something about my model, I might have to change evaluation code, too.
2. Support an optional validation dataset during traing and try to make logging of quantities seamless, i.e. we should then simply log everything we log on the training data also on the validation data. However, I think it is crucial to make sure that this does not introduce coding overhead. This would probably require us to integrate logging more closely into the base iterator which could be good on its own.
3. Keep training and evaluation seperate but reduce the overhead associated with evaluation. For example, we could try to design a Training Iterator base class which can be reused as an evaluator by simply skipping the optimization. That way, we would also track the same quantities as on the train set (that part is actually overlapping with the previous point). To then support more frequent evaluations, we could implement the ability to send the training process a signal to write a checkpoint - an evaluator could then basically request a temporary checkpoint which it can use to evaluate the model and then immediately delete if it wants to. In fact, we already have something similar in place, SIGINT writes a checkpoint (but then terminates the training process, too).

I kind of like the third option, but the second one also seems reasonable. What do you/others think?

**Comment by LeanderK** [comment]
Thanks for your answer!

> As for switching halfway to a different kind of dataset, I would definitely suggest to split the training into two stages and just resume training on the new dataset in the second stage. Is that feasible or do you think there is a really good reason to change the dataset during training?

I think it's feasible, although maybe a bit inelegant. It will probably show whether it turns out to be a real problem (I don't think so...).

> Collections of datasets as you mention: This would enable different datasets in a single process, but the only use case I can see is that of evaluating the model. If we go that route, then we basically abandon the idea of separation between training and evaluation and we always have to handle logic for training and evaluation during development. I actually prefer the separation because it allows quicker iteration: I can first write whatever code is necessary to start my training, and then I can still think about evaluation code while it trains. If we mix training and evaluation and I change something about my model, I might have to change evaluation code, too.

I think this might be a source of my problem with edflow in it's current state. I need to evaluate every epoch, I am mainly interested in these statistics on my test-dataset and my evaluation code is by code-design independent of my concrete model (provided I pass the right dataset). I don't really have to think much about evaluation and mainly about the model. Also, my evaluation metrics are quite complicated. I track stuff like numerical instability of certain layers, different histograms of latents I am interested in, etc. I have 3 types of experiments that are essentially fixed and independent of the dataset provided I have the right model to train. I would like to have an edflow-abstraction so that I can choose the right experiments and I get all the relevant dataset-versions and evaluations for these experiments.

The code that pieces together the right evaluations (metrics etc.) is approx. 500 lines of python and the code that handles the datasets is also quite enourmous. When I try to classify with my model I have different stats that I want to track compared to density estimation. I really don't want to do this via some .yaml flags (I would probably generate them via another python script, which seems highly suboptimal), but instead, have code that provides me per default with the right evaluation strategies.

> Keep training and evaluation separate but reduce the overhead associated with evaluation. (...) . That way, we would also track the same quantities as on the train set (that part is actually overlapping with the previous point) (...)

I don't really track most of these metrics in the training because they are computationally intensive. In fact, most of these metrics are only evaluated every 10 epochs to not slow down the training.

Maybe rethinking the data-abstraction is not the right approach, but I need something to piece together some common things for the 3 basic types of experiments I am doing. I have a lot of evaluations I am interested in, some overlap between these types, some not. It might be that edflow, although certainly interesting, is not really made for my use-case (since I have quite complicated metrics for my model with many depending on entirely different or only slightly different versions of the dataset).
 
To summarize: I have quite a huge project with 2 sources of complexity I am not satisfied by. My code that pieces together the datasets is quite huge and the code that pieces together the the "type" of the experiment. A more modular apprach would be nice, maybe rethinking the trainer abstraction would be more approriate. But it also may be that edflow is simply not made for my use-case. This would be dissapointing because I think it could fit quite nicely.

**Comment by pesser** [comment]
Sorry I'm not really able to help, but I'd be happy if you have some suggestions to improve the workflow with edflow! I'll try to get some feedback regarding points 1-3 of my previous comment during the next meeting.

**Comment by pesser** [comment]
From our meeting: We need more documentation, but we discussed the three original points and came to the conclusion that

1. every dataset can be a training or testing dataset. it can even decide automatically based on the config key `test_mode` which is set to true for evaluation configs (started with -e)
2. We tried our best to keep the design of the dataset class as minimal as possible to remain as flexible as possible and it is not clear in what sense they are not flexible enough or how they could be made more flexible. Changing datasets during training is not something we want to support and if really necessary it can be achieved with a dataset that returns examples from both datasets and the iterator or model pick only examples from the first dataset during the first training phase and then only examples from the second dataset during the second training phase.
3. there is support for complex datasets. For multi-task training, datasets can be concatenated or joined with one of the helpers in dataset.py or a new dataset can instantiate all the sub datasets and return examples of each as desired in its get_example method. Different datasets for different evaluations are explicitly supported because each evaluation config can specify its own dataset.

Thus, I'm closing for now.
